File: python/pid.py
from __future__ import annotations

from dataclasses import dataclass


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


@dataclass
class PID:

    kp: float
    ki: float = 0.0
    kd: float = 0.0
    output_limits: tuple[float, float] = (-1.0, 1.0)
    integral_limits: tuple[float, float] = (-100.0, 100.0)
    integral_decay: float = 0.0

    integral: float = 0.0
    previous_error: float | None = None
    previous_output: float = 0.0

    def reset(self) -> None:
        self.integral = 0.0
        self.previous_error = None
        self.previous_output = 0.0

    def update(self, error: float, dt: float) -> float:
        if dt <= 0.0:
            return self.previous_output

        derivative = 0.0
        if self.previous_error is not None:
            derivative = (error - self.previous_error) / dt

        candidate_integral = clamp(
            self.integral*self.integral_decay + error * dt, # утечка интеграла против wind-up
            self.integral_limits[0],
            self.integral_limits[1],
        )
        raw_output = (
            self.kp * error
            + self.ki * candidate_integral
            + self.kd * derivative
        )

        low, high = self.output_limits
        saturated_high_in_error_direction = raw_output > high and error > 0.0 #проверка на упор в лимиты
        saturated_low_in_error_direction = raw_output < low and error < 0.0
        if not (
            saturated_high_in_error_direction
            or saturated_low_in_error_direction
        ):
            self.integral = candidate_integral

        output = clamp(
            self.kp * error + self.ki * self.integral + self.kd * derivative,
            low,
            high,
        )
        self.previous_error = error
        self.previous_output = output
        return output

File: arduino/autopilot/pid.cpp
#include "pid.h"
float clampFloat(float value, float low, float high) {
  if (value < low) return low;
  if (value > high) return high;
  return value;
}

SimplePID::SimplePID() {
  begin(0.0f, 0.0f, 0.0f, -1.0f, 1.0f, -100.0f, 100.0f);
}

void SimplePID::begin(float kp, float ki, float kd, float outMin, float outMax,
                      float integralMin, float integralMax) {
  kp_ = kp;
  ki_ = ki;
  kd_ = kd;
  outMin_ = outMin;
  outMax_ = outMax;
  integralMin_ = integralMin;
  integralMax_ = integralMax;
  reset();
}

void SimplePID::reset() {
  integral_ = 0.0f;
  previousError_ = 0.0f;
  previousOutput_ = 0.0f;
  hasPreviousError_ = false;
}

float SimplePID::update(float error, float dt) {
  if (dt <= 0.0f) return previousOutput_;

  float derivative = 0.0f;
  if (hasPreviousError_) {
    derivative = (error - previousError_) / dt;
  }

  float candidateIntegral = clampFloat(
    integral_*0.999 + error * dt, // добавил утечку в интеграл, чтобы бороться с windup
    integralMin_,
    integralMax_
  );

  

  float rawOutput = kp_ * error + ki_ * candidateIntegral + kd_ * derivative;
  bool saturatedHighInErrorDirection = rawOutput > outMax_ && error > 0.0f;
  bool saturatedLowInErrorDirection = rawOutput < outMin_ && error < 0.0f;

  if (!saturatedHighInErrorDirection && !saturatedLowInErrorDirection) {
    integral_ = candidateIntegral;
  }

  rawOutput = kp_ * error + ki_ * integral_ + kd_ * derivative;
  float output = clampFloat(rawOutput, outMin_, outMax_);

  previousError_ = error;
  previousOutput_ = output;
  hasPreviousError_ = true;
  return output;
}
